---
title: "R Notebook"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---
```{r}

knitr::opts_chunk$set(
  warning = FALSE,
  error = TRUE,
  echo = TRUE
)
```

```{r setup}
library("ggbiplot")
library("data.table")
library("tidyverse")
library("readxl")
library("glue")
library("gggenes")
library("ggtext")
library("broom")
library("skimr")
library("here")
setwd(here())
```

## Goal: Do the operon identification in pipes using dplyr
## Date initiated: 23 maj 2021
## Author: Soren H

Defining parameter that are used in the filtering of operons later. The script support processing fo multiple queries at one time. 

```{r}
# The name of the .txt files
filename_psiblast <- c("pnag_ica")
# Percent identity cutoff (> not >=)
perc_id = 20
# Max prokka number difference netween operon genes
max_dist_prok = 8
# Max nucleotide difference between genes
max_dist_gene = 2000
# Minimum genes in operons
min_genes = 2
# Number of genes on each side in figure
flanking_genes = 2
```



```{r}
# Statistics on the HQ-MAGs from singleton et.al 2021
magstats <- fread("./data/raw/magstats.tsv", sep = "\t")

# File with metadata on the query genes, e.g. function
query_metadata <- excel_sheets("./data/raw/Query_figur.xlsx") %>%
  sapply(function(X) read_xlsx("./data/raw/Query_figur.xlsx", sheet = X, skip = 1), USE.NAMES = T) %>% 
  lapply(as.data.frame) %>% 
  subset(!(names(.) %in% c("Abbreveations", "HA_S_pyogenes"))) %>%
  rbindlist()

# Additional information of the prokka annotations
gff <- fread("./data/raw/gff.tsv") %>%
  mutate(Target_label = paste(ID, ProkkaNO, sep = "_"),
         ProkkaNO = as.numeric(ProkkaNO)) %>%
  filter(!is.na(ProkkaNO))

# Manually verified operons
verified <- glue("./data/processed/manually_verified/{filename_psiblast}.txt") %>% 
  file.info() %>% 
  filter(size > 0) %>% 
  rownames() %>% 
  lapply(fread, header = FALSE) %>% 
  rbindlist() %>% 
  unique() %>% 
  `[[`(1)
  
  
```


```{r}
### Loading raw psiblast results
test <- filename_psiblast %>%
  paste0("./data/raw/psiblast/", ., ".txt") %>%
  lapply(
    fread,
    header = FALSE, sep = "\t", fill = TRUE, 
    col.names = c(
      "Query_label", "Target_label", "Percent_identity", "align_length",
      "mismatch", "gap_opens", "start_query", "end_query",
      "start_target", "end_target", "E_value", "Bit score")
  ) %>%
  bind_rows() %>%
# ->ASSIGNING psiblast_raw<- 
  assign(x = "psiblast_raw", value = ., pos = 1) %>% 
  ### Processesing psiblast results
  ## Filtering
  subset(!is.na(E_value)) %>%
  group_by(Target_label) %>%
  filter(`Bit score` == max(`Bit score`)) %>%
  filter(!duplicated(Target_label)) %>%
  filter(Percent_identity > perc_id) %>%
  ## Cleaning
  separate(Target_label, c("ID", "ProkkaNO"),
           sep = -5, remove = FALSE, extra = "merge") %>%
  mutate(ProkkaNO = as.numeric(ProkkaNO)) %>%
  mutate(ID = str_sub(ID, start = 1, end = -2)) %>% 
  ## Merging
  left_join(gff,
            by = c("Target_label", "ProkkaNO", "ID"),
            keep = FALSE) %>%
  left_join(magstats, by = "ID", keep = FALSE) %>%
  left_join(query_metadata[, c("Genename", "Function")], by = c("Query_label" = "Genename"))  %>%
  arrange(Target_label, ProkkaNO) %>%
  ## Operon Grouping
  group_by(seqname, ID) %>%
  mutate(prio_prok = (ProkkaNO - shift(ProkkaNO, 1)) %>%  replace_na(max_dist_gene + 1),
         post_prok = (shift(ProkkaNO, -1)) - ProkkaNO %>%  replace_na(max_dist_gene + 1),
         prio_gene = (start - shift(end, 1)) %>%  replace_na(max_dist_gene + 1),
         post_gene = (shift(start, -1) - end) %>% replace_na(max_dist_gene + 1)) %>%
# ->ASSIGNING psiblast_filtered<- 
  assign(x = "psiblast_filtered", value = ., pos = 1) %>% 
  # Remove those who do not satisfy any of the distance requirements
  filter(prio_prok < max_dist_prok | post_prok < max_dist_prok |
         prio_gene < max_dist_gene | post_gene < max_dist_gene) %>%
  ungroup() %>%
  # If a gene only satisfy distance to posterior hit and not prior, it must be operon start
  mutate(operon_place = ifelse((post_prok < max_dist_prok | post_gene < max_dist_gene) &
                              !(prio_prok < max_dist_prok | prio_gene < max_dist_gene),
                              yes = "start", no = "."),
         operon = ifelse(operon_place == "start", row_number(), NA)) %>%
  fill(operon, .direction = "down") %>%
  ## Operon filtering
  group_by(operon) %>%
  filter(length(unique(Query_label)) >= min_genes) %>%
  select(-prio_prok, -prio_gene, -post_prok, -post_gene) %>% 
  ## 
  ungroup() %>% 
  nest(cols = !c(ID, operon)) %>% 
  group_by(ID) %>% 
  mutate(letter = toupper(letters)[seq_along(unique(operon))],
         ID2 = case_when(
           length(unique(operon)) > 1 ~ paste(ID, letter, sep = "_"),
           TRUE ~ ID)
         ) %>% 
  select(-letter) %>% unnest(cols = c(cols)) %>% 
# ->ASSIGNING "psiblast"<- 
  assign(x = "psiblast", value = ., pos = 1) %>% 
  ### Expansion of "psiblast" with surrounding genes
  ## Identify surrounding genes
  group_by(operon) %>%
  fill(ID2, .direction = "updown") %>% 
  summarise(
    ID = unique(ID),
    tig = unique(seqname),
    max = max(ProkkaNO),
    min = min(ProkkaNO)
    ) %>%
  full_join(gff) %>%
  group_by(operon) %>%
  filter(ProkkaNO <= (max + flanking_genes) &
         ProkkaNO >= (min - flanking_genes) & 
         seqname  == tig) %>%
  ## Merging surrounding and psiblast genes
  full_join(psiblast) %>%
  ## Prepare for plotting
  group_by(operon) %>%
  mutate(
    # gggenes require 1 and -1 as strand direction
    strand = case_when(strand == "+" ~ 1, strand == "-" ~ -1, TRUE ~ as.numeric(strand)),
    # Relative distances (start always minimum value of gene posistion)
    end = end - min(start),
    start = start - min(start),
    # converting AA numbers to nucleotide numbers
    end_target_plot =   ifelse(is.na(end_target),   end,   end_target   * 3 + start),
    start_target_plot = ifelse(is.na(start_target), start, start_target * 3 + start),
    # Assigning artificial percent identity to domains
    Percent_identity = ifelse(is.na(Percent_identity), 40, Percent_identity),
    # Order operons from highest bit score to lowest
    operon = factor(operon, levels = unique(arrange(., -`Bit score`)$operon))
    ) %>%
  fill(MiDAS3_6Tax, .direction = "updown") %>%
  fill(SILVA138Tax, .direction = "updown") %>%
  fill(GTDBTax, .direction = "updown") %>%
  fill(ID2, .direction = "updown") %>% 
# ->ASSIGNING "genes"<- 
  assign(x = "genes", value = ., pos = 1)
```






```{r}
### Loading interproscan data and making domain data
  ## Loading
filename_psiblast %>% 
  lapply(function(query) {
    # Use this function on each polysaccharide name (e.g. "cellulose1)
    list.files(paste0("./data/raw/ips/", query, "_fasta_removed/")) %>%
      subset(grepl(., pattern = query)) %>% 
      lapply(function(mag) {
        # Use this function on each mag ID (e.g. "Ega_18-Q3-R5-49_MAXAC.199")
        mag_path = paste0("./data/raw/ips/", query, "_fasta_removed/", mag)
        if (file.info(mag_path)$size > 0) {
          read.table(
            mag_path,
            sep = "\t",
            fill = TRUE,
            comment.char = "#"
          )
        }
      }) %>%
      bind_rows()
    }) %>%
  bind_rows() %>%
  ## Extraction of relevant annotation information
  filter(!str_detect("polypeptide", V3)) %>%
  mutate(
    domain = str_extract(V9, "signature_desc=[^;]*;"),
    domain = str_sub(domain, 1, -2),
    domain = gsub("signature_desc=", "", x = domain)
    ) %>%
  subset(select = -c(V2, V3, V7, V8, V9)) %>%
  setNames(c("Target_label", "start1", "end1", "e_value", "Domain")) %>%
  ## Combining information from genes to domains
  full_join(genes, .) %>%
  mutate(start2 = start + as.numeric(start1) * 3,
         end2 = start + as.numeric(end1) * 3) %>%
  subset(select = c(
      "start", "end", "start2", "end2", "Domain", "operon", "ID",
      "ID2", "Function", "strand")) %>%
  mutate(Percent_identity = 50) %>%
  filter(!is.na(Domain)) %>%
  distinct() %>%
  ## Formating of domain names (a bit of confusing regular expressions)
  mutate(
    Domain = str_replace(Domain, "[gG]lycosyl.*transferase.*[fF]amily ", "GT f"),
    Domain = str_replace(Domain, "[gG]lycosyl.*transferase.*[gG]roup ", "GT g"),
    Domain = str_replace(Domain, "[gG]lycosyl.*transferase.[lL]ike.[fF]amily", "GT like f"),
    Domain = str_replace(Domain, "[gG]lycosyl.*transferase", "GT"),
    Domain = str_replace(Domain, "[gG]lycosyl [hH]ydrolase.*[fF]amily " , "GH"),
    Domain = str_replace(Domain, "[gG]lycosyl [hH]ydrolase" , "GH"),
    Domain = str_replace(Domain, ".*[cC]ellulose.*synth.*protein[^' ']*" , "CS "),
    Domain = str_replace(Domain, ".*[cC]ellulose.*synth[^' ']*", "CS "),
    Domain = str_replace(Domain, " N.terminal domain" , "_N"),
    Domain = str_replace(Domain, " C.terminal domain" , "_C"),
    Domain = str_replace(Domain, ".*BCSC_C.*", "bcsC"),
    Domain = str_replace(Domain, "[iI]nitation [fF]actor" , "IF"),
    Domain = str_replace(Domain, "[eE]longation [fF]actor" , "EF"),
    Domain = str_replace(Domain, ".*Tetratrico.*|.*TPR.*", "T"),
    Domain = str_replace(Domain, ".*[dD]omain of unknown function.*", "NA"),
    ) %>% 
# -> ASSIGNING "domains"
  assign(x = "domains", value = ., pos = 1)
```

Plotting using GGgenes

```{r}
#### Adding midas taxonomy names and modify appearance to be more neat 
# e.g. Candidatus in cursive instead of Ca_
add_midas_tax <- function(data) {
  genes %>% 
    select(ID2, MiDAS3_6Tax) %>% 
    separate(MiDAS3_6Tax, into = c("drop", "MiDAS3_6Tax"), sep = "=") %>% 
    select(-drop) %>% 
    mutate(
      MiDAS3_6Tax = str_remove_all(MiDAS3_6Tax, ".[\\:\\;]")
    ) %>% 
    separate(MiDAS3_6Tax, into = c("mi_domain","mi_phylum", "mi_class", "mi_order", "mi_family", "mi_genus", "mi_species"), sep = ",") %>% 
    distinct() %>% 
    right_join(data) %>% 
    mutate(
      mi_species = str_remove(mi_species, paste0(mi_genus, "_")),
      title = paste0(ID2, "<br>",
      "", mi_phylum, "<br>",
      "", mi_class, "<br>",
      "", mi_order, "<br>",
      "", mi_family, "<br>",
      "*", mi_genus, "*", "<br>",
      "*", mi_species, "*"),
      # Formating taxa names to be more inline with recommended guidelines
      title = str_replace_all(title, 
                              pattern = "\\*Ca_([^*]*)\\*", 
                              replacement = "*Candidatus* \\1"),
      title = str_replace_all(title, 
                              pattern = "\\Ca_(.*) ", 
                              replacement = "Candidatus \\1"),
      title = str_replace_all(title, 
                              pattern = "\\*(.*)\\_marine\\_group\\*", 
                              replacement = "\\1 marine group"),
      title = str_replace_all(title, 
                              pattern = "\\*(.*)\\_marine\\_group\\*", 
                              replacement = "\\1 marine group"),
      title = str_replace_all(title, 
                              pattern = "_Subgroup_(.)", 
                              replacement = " (Subgroup \\1)"),
    )
}
genes <- add_midas_tax(genes)
domains <- add_midas_tax(domains)




gene_height <- 4
ggsave(glue("./figures/gggenes_", paste(filename_psiblast, collapse = "_"), "_min_genes{min_genes}.pdf"), 
       width = unit(13, "mm"),
       height = unit(1.5*length(unique(genes$ID2)), "mm"),
       limitsize = FALSE,
ggplot(genes, aes(xmin = start, xmax = end, y = title, forward = strand)) +
  # Empty gene arrows
  geom_gene_arrow(
    arrowhead_height = unit(gene_height, "mm"),
    arrow_body_height = unit(gene_height, "mm"),
    arrowhead_width = unit(5, "mm")
  ) +
  # Colored gene arrows (match in psiblast)
  geom_subgene_arrow(
    data = genes,
    mapping = aes(xmin = start, xmax = end, y = title,
      xsubmin = start_target_plot,
      xsubmax = end_target_plot,
      fill = Function,
      forward = strand,
      alpha = Percent_identity
    ),
    arrowhead_height = unit(gene_height, "mm"),
    arrow_body_height = unit(gene_height, "mm"),
    arrowhead_width = unit(5, "mm"),
    #position = position_nudge(y = 0.3)
  ) +
  geom_text(data = genes %>% mutate(start = (start_target_plot + end_target_plot)/2),
            aes(x = start, label = Query_label)) +
  geom_text(data = genes %>% mutate(Percent_identity = ifelse(Percent_identity == 40,
                                                              yes = " ",
                                                              no = paste0(
                                                                signif(Percent_identity, digits = 2),
                                                                "%"))),
            aes(x = start_target_plot, label = Percent_identity),
            nudge_x = 200, nudge_y = 0.16, size = 2) +
  # Domains boxes
  geom_gene_arrow(
    data = domains,
    mapping = aes(
      xmin = start2,
      xmax = end2,
      y = title,
      forward = strand,
      fill = Function,
      alpha = Percent_identity
    ),
    arrowhead_height = unit(gene_height - 1, "mm"),
    arrow_body_height = unit(gene_height - 1, "mm"),
    arrowhead_width = unit(0, "mm"),
    position = position_nudge(y = -0.35)
  ) +
  geom_text(
    data = domains %>% mutate(start = (start2 + end2) / 2),
    aes(x = start, label = Domain,  y = title),
    nudge_y = -0.35,
    size = 2,
    angle = 8
  ) +
  facet_wrap( ~ operon, scales = "free", ncol = 1) +
  scale_alpha_continuous(range = c(0.1, 1), limits = c(20, 40)) +
  guides(alpha = guide_legend(override.aes = list(fill = "black"))) +
  theme_genes() +
  theme(
    legend.position = "top",
    axis.text.y = element_markdown(),
    axis.title.y = element_blank()
  ) +
  scale_fill_brewer(palette = "Set3")
)


```



Exploratory plots

Evaluation of the change in bit score distrubution and manual verification. 
```{r}
psiblast_compare <- psiblast_filtered %>% 
  mutate(
    in_operon = case_when(
      Target_label %in% subset(psiblast, ID %in% verified)$Target_label ~ "Manual",
      Target_label %in% psiblast$Target_label ~ "Automatic",
      TRUE ~ "Removed"
    ),
    length = (end - start + 1) / 3,
    target_qov =  (end_target - start_target) / length,
    bit_identity_qov = Percent_identity * `Bit score` * target_qov
  )


ggsave(filename = paste0("./figures/score_dist_", paste(filename_psiblast, collapse = "_"),".pdf"), 
ggplot(psiblast_compare, aes( x = Query_label, y = bit_identity_qov, col = in_operon)) +
  geom_boxplot(outlier.colour = NA) +
  geom_jitter(alpha = 0.4, size = 0.7, position = position_jitterdodge(jitter.width = 0.25)) +
  theme_bw() +
  theme(panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank()
  ) #+
  #scale_y_log10()
)
```










```{r}
library(ggbiplot)
psiblast_compare %>% 
  filter(!is.na(E_value)) %>% 
  ungroup() %>% 
  mutate(coverage = (end_target - start_target) / (end - start)/3,) %>% 
  select(`Bit score`, Percent_identity, coverage) %>% 
  prcomp(center = TRUE, scale. = TRUE) %>% 
  ggbiplot(groups = psiblast_compare$in_operon, ellipse = TRUE)

psiblast_PCA_mean <-  psiblast %>% 
  ungroup() %>% 
  mutate(coverage = (end_target - start_target) / (end - start)/3,) %>%
  group_by(operon) %>% 
  dplyr::summarize(
    mean_bit = mean(`Bit score`),
    mean_cov = mean(coverage),
    mean_perc_id = mean(Percent_identity),
    size = length(coverage)
    ) %>%
  inner_join(psiblast, by = "operon") %>% 
  ungroup() %>% 
  full_join(psiblast_filtered) %>%
  mutate(coverage = (end_target - start_target) / (end - start)/3,) %>%
  mutate(
    in_operon = case_when(
      Target_label %in% subset(psiblast, ID %in% verified)$Target_label ~ "Manual",
      Target_label %in% psiblast$Target_label ~ "Automatic",
      TRUE ~ "Removed"
    ),
    mean_bit = ifelse(is.na(mean_bit), yes = `Bit score`, no = mean_bit), 
    mean_cov = ifelse(is.na(mean_cov), yes = coverage, no = mean_cov),
    mean_perc_id = ifelse(is.na(mean_perc_id), yes = Percent_identity, no = mean_perc_id),
    size = ifelse(is.na(size), 1, size),
    size2 = as.character(size)
    ) %>%
  ungroup()
psiblast_PCA_mean2 <- psiblast_PCA_mean %>%  
 select(mean_bit, mean_cov, mean_perc_id, size) 


prcomp(psiblast_PCA_mean2, scale. = TRUE, center = TRUE) %>% 
  ggbiplot(groups = psiblast_PCA_mean$size2)

```


```{r}
C <- chol(S <- toeplitz(.9 ^ (0:31))) # Cov.matrix and its root
all.equal(S, crossprod(C))
set.seed(17)
X <- matrix(rnorm(32000), 1000, 32)
Z <- X %*% C  ## ==>  cov(Z) ~=  C'C = S
all.equal(cov(Z), S, tolerance = 0.08)
pZ <- prcomp(Z, tol = 0.1)
```
















