---
title: "Figure 4: Primer bias V13 vs V4"
author: "Morten Simonsen Dueholm"
date: "2021-08-25"
---

#Load packages
```{r load_packages, message=FALSE, warning=FALSE, echo=FALSE, include=FALSE}
library(ampvis2)
library(data.table)
library(tidyverse)
library(patchwork)
```

#Load MD data
```{r MD data}
#load metadata
seq_metadata <- read.csv("data/2020-07-03_Sequencing_metadata.txt", sep="\t")
wwtp_m <- read.csv("data/DataS1_210413.txt", sep="\t")
wwtp_m$Mean <- "mean"
wwtp_m$abs_Latitude <- abs(wwtp_m$Latitude)

#V13
seq_V13 <- seq_metadata[ , which(names(seq_metadata) %in% c("V13_seq_id","WWTP_id"))]

V13metadata <- merge.data.frame(seq_V13, wwtp_m, by ="WWTP_id", all.x = TRUE)
V13metadata$WWTP_ID <- V13metadata$WWTP_id
V13metadata <- V13metadata[,-1]

d13 <- amp_load(otutab = "data/V13_ASVtab.txt", 
                taxonomy = "data/V13ASV_vs_MiDAS_4.8.sintax",
                metadata = V13metadata)

# Remove samples with low read count
d13n <- amp_subset_samples(d13, minreads = 10000, normalise = TRUE)

#Subset AS only in 4 basic process_type(s)
d13nAS <- amp_subset_samples(d13n, Plant_type == "Activated sludge")
d13nAS4PT <-  amp_subset_samples(d13nAS, Process_type %in% c("C", "C,N", "C,N,DN", "C,N,DN,P"))  

##V4
seq_V4 <- seq_metadata[ , which(names(seq_metadata) %in% c("V4_seq_id","WWTP_id"))]
V4metadata <- merge.data.frame(seq_V4, wwtp_m, by ="WWTP_id", all.x = TRUE)
V4metadata$WWTP_ID <- V4metadata$WWTP_id
V4metadata <- V4metadata[,-1]

d4 <- amp_load(otutab = "data/V4_ASVtab.txt", 
               taxonomy = "data/V4ASV_vs_MiDAS_4.8.sintax", 
               metadata = V4metadata)

# Remove samples with low read count
d4n <- amp_subset_samples(d4, minreads = 10000, normalise = TRUE)

#Subset AS only in 4 basic process_type(s)
d4nAS <- amp_subset_samples(d4n, Plant_type == "Activated sludge")
d4nAS4PT <-  amp_subset_samples(d4nAS, Process_type %in% c("C", "C,N", "C,N,DN", "C,N,DN,P"))  
```

#V13-V4 genus abundance
```{r V13 V4 genus abundance}
###mean abundance V13
#remove ASVs with no Genus classification
ASVs_nogenus <- as.character(d13nAS4PT$tax[d13nAS4PT$tax$Genus == "" | is.na(d13nAS4PT$tax$Genus),"OTU"])
d13nAS4PT_clas_g <- amp_subset_taxa(d13nAS4PT, tax_vector = ASVs_nogenus, remove = TRUE)

gV13 <- amp_heatmap(d13nAS4PT_clas_g,
            group_by = c("Mean"),
            tax_aggregate = "Genus",
            measure = "median",
            tax_show = 7000,
            normalise = FALSE,
            textmap = TRUE
            )

gV13 <- setDT(gV13, keep.rownames = TRUE)
gV13 <- as.data.frame(gV13)
names(gV13)[1] <- "Genus"
names(gV13)[2] <- "V13_mean_abundance"

###mean abundance V4
#remove ASVs with no Genus classification
ASVs_nogenus <- as.character(d4nAS4PT$tax[d4nAS4PT$tax$Genus == "" | is.na(d4nAS4PT$tax$Genus),"OTU"])
d4nAS4PT_clas_g <- amp_subset_taxa(d4nAS4PT, tax_vector = ASVs_nogenus, remove = TRUE)

gV4 <- amp_heatmap(d4nAS4PT_clas_g,
            group_by = c("Mean"),
            tax_aggregate = "Genus",
            measure = "median",
            tax_show = 7000,
            normalise = FALSE,
            textmap = TRUE
            )

gV4 <- setDT(gV4, keep.rownames = TRUE)
gV4 <- as.data.frame(gV4)
names(gV4)[1] <- "Genus"
names(gV4)[2] <- "V4_mean_abundance"

#merge V13 and V4 data
gV134 <- merge.data.frame(gV13, gV4, by ="Genus", all.x = TRUE, all.y = TRUE)

#replace NA's with 0
gV134[is.na(gV134)] <- 0

#remove Genera that are < 0.001% abundant in either V13 or V4 
gV134 <- gV134 %>% filter(V13_mean_abundance >= 0.001 | V4_mean_abundance >= 0.001)

#Change all values <0.001 to 0.001 to allow calculation of fold difference in between V13 and V4 
gV134[gV134<0.001] <- 0.001

#Calculate -log2(fold change) (FC)
gV134 <- gV134 %>% 
  mutate(FC=-log2(V13_mean_abundance/V4_mean_abundance))

#Subset genera based on their preferential detection
gV134 <- gV134 %>%
  mutate(Bias=if_else(FC>1,"More abundant with V4",if_else(FC<(-1), "More abundant with V1-V3", "Equally detected")))

summary <- gV134 %>%
  group_by(Bias) %>%
  summarise(Counts = n())

write.csv(gV134,"output/DataS2.csv", row.names = FALSE, quote = FALSE)

#Create scatter plot
p <- ggplot(gV134, aes_string(x = "V13_mean_abundance", y= "V4_mean_abundance", color="Bias")) + 
  geom_point(size = 2) +
  geom_abline(size = 0.5) +
  geom_text(data = filter(gV134, gV134$FC>1 & gV134$V4_mean_abundance>0.1), aes(label = Genus), vjust = 1.5, colour="grey30", size = 3) +
  geom_text(data = filter(gV134, gV134$FC<(-1) & gV134$V13_mean_abundance>0.1), aes(label = Genus), vjust = 1.5, colour="grey30", size = 3) +            
  xlab("Mean V13 amplicon genus read abundance (%)") + 
  ylab("Mean V4 amplicon genus read abundance (%)") +
  scale_x_log10(limits=c(0.001,2)) +
  scale_y_log10(limits=c(0.001,2)) +
  theme_bw() +
  theme(legend.position = "none")

ggsave(filename="output/Figure4a_V13_vs_V4_genus_abundance.pdf", plot=p, width=6, height=4, useDingbats=FALSE, limitsize=FALSE)

#Create heatmaps
gV134_V13 <- gV134 %>%
  filter(FC<(-1)) %>%
  arrange(desc(V13_mean_abundance)) %>%
  select(1:3) %>%
  slice(1:20) %>%
  rename("V1-V3" = V13_mean_abundance, "V4" = V4_mean_abundance) %>%
  gather(2:3, key="Amplicon",value = "Abundance")

gV134_V13sub <- gV134_V13 %>% filter(Amplicon=="V1-V3")
gV134_V13$Genus <- reorder(gV134_V13sub$Genus, gV134_V13sub$Abundance)  

gV134_V4 <- gV134 %>%
  filter(FC>1) %>%
  arrange(desc(V4_mean_abundance)) %>%
  select(1:3) %>%
  slice(1:20) %>%
  rename("V1-V3" = V13_mean_abundance, "V4" = V4_mean_abundance) %>%
  gather(2:3, key="Amplicon",value = "Abundance")

gV134_V4sub <- gV134_V4 %>% filter(Amplicon=="V4")
gV134_V4$Genus <- reorder(gV134_V4sub$Genus, gV134_V4sub$Abundance)

p1 <- ggplot(gV134_V13, aes(Amplicon, Genus)) +
    geom_tile(aes(fill = Abundance)) + 
    geom_text(aes(label = round(Abundance, 2))) +
    scale_fill_gradientn(colors = rev(c("#a50026", "#d73027", "#f46d43", "#fdae61",
                                   "#fee090","#ffffbf","#e0f3f8","#abd9e9","#74add1",
                                   "#4575b4","#313695")), trans = "log10",, breaks = c(0.001, 0.01, 0.1, 1), limits=c(0.001,1)) +
  #scale_fill_viridis(option="plasma", trans = "log10") +
    theme_bw()

p2 <- ggplot(gV134_V4, aes(Amplicon, Genus)) +
    geom_tile(aes(fill = Abundance)) + 
    geom_text(aes(label = round(Abundance, 2))) +
    scale_fill_gradientn(colors = rev(c("#a50026", "#d73027", "#f46d43", "#fdae61",
                                   "#fee090","#ffffbf","#e0f3f8","#abd9e9","#74add1",
                                   "#4575b4","#313695")), trans = "log10",, breaks = c(0.001, 0.01, 0.1, 1),
                         limits=c(0.001,1)) +
  theme_bw()

p3 <- p1+p2

ggsave(filename="output/Figure4b_V13_vs_V4.pdf", plot=p3, width=8, height=5, useDingbats=FALSE, limitsize=FALSE)
```
