---
title: "Figure 5c: Phylogenetic trees for primerbias"
author: "cko"
date: "2021-09-01"
---
  
  #Load packages
```{r load_packages, message=FALSE, warning=FALSE, echo=FALSE, include=FALSE}
library(tidyverse)
library(ggtree)
library(patchwork)
```

#Load data
```{r, echo=FALSE, message=FALSE, warning=FALSE}
#load metadata
tax_metadata <- read.csv("data/tax_complete.csv", sep=",", header = TRUE)

#load tree
tree_bacteria <- read.tree(file = "data/bacteria_newick2.tree")

# Read primer mapping results
read_primerbias <- function(input) {
  map <- read.delim(input,
                      sep = ",",
                      header = FALSE,
                      quote = "\"",
                      fill = TRUE,
                      check.names = FALSE,
                      stringsAsFactors = FALSE,
                      skip=12) %>%
    select(1,10) %>%
    rename(FLASV = "V1") %>%
    arrange(readr::parse_number(FLASV)) %>%
    mutate(p_id=gsub("data/primerbias/(.+)_FLASVs_hits.txt", "\\1", paste(input))) %>%
    separate(p_id,into=c("Primer","Primer_set"), sep="_") %>%
    mutate(V10=as.numeric(V10))}

file_list <- list.files("data/primerbias/", full.names=TRUE)

for (file in file_list){
  
# if the merged dataset doesn't exist, create it

  if (!exists("dataset")){
    dataset <- read_primerbias(file)
  }
  
  # if the merged dataset does exist, append to it

  if (exists("dataset")){
    temp_dataset <-read_primerbias(file)
    dataset<-rbind(dataset, temp_dataset)
    rm(temp_dataset)
  } 

}
```

## Summarize_data
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
Summary <- dataset %>%
  group_by(FLASV,Primer_set) %>%
  summarize("Match"=if_else(max(V10)==0, "Perfect", if_else(max(V10)>1, "Poor", "Partial"))) %>%
  spread("Primer_set", "Match")
```

#Create metadata for bacterial and archaeal SeqID trees
```{r, echo=FALSE, message=FALSE, warning=FALSE}
Perfect <- (Summary %>%
  filter(bac10=="Perfect") %>%
  select(FLASV))$FLASV

Partial <- (Summary %>%
  filter(bac10=="Partial") %>%
  select(FLASV))$FLASV

Poor <- (Summary %>%
  filter(bac10=="Poor") %>%
  select(FLASV))$FLASV

groupInfo1 <- list(Perfect = Perfect, 
                   Partial = Partial, 
                   Poor = Poor)
```


#Create metadata for bacterial Taxonomy tree (Important taxa)
```{r, echo=FALSE, message=FALSE, warning=FALSE}
Sphingobacteriales <- (tax_metadata %>%
  filter(grepl("Sphingobacteriales",Order)) %>%
  select(FLASV))$FLASV

Cytophagales  <- (tax_metadata %>%
  filter(grepl("Cytophagales",Order)) %>%
  select(FLASV))$FLASV

Acidobacteriota <- (tax_metadata %>%
  filter(grepl("Acidobacteriota",Phylum)) %>%
  select(FLASV))$FLASV

Myxococcota <- (tax_metadata %>%
  filter(grepl("Myxococcota",Phylum)) %>%
  select(FLASV))$FLASV

Chloroflexi <- (tax_metadata %>%
  filter(grepl("Chloroflexi",Phylum)) %>%
  select(FLASV))$FLASV

Cyanobacteria <- (tax_metadata %>%
  filter(grepl("Cyanobacteria",Phylum)) %>%
  select(FLASV))$FLASV

Planctomycetota <- (tax_metadata %>%
  filter(grepl("Planctomycetota",Phylum)) %>%
  select(FLASV))$FLASV

Patescibacteria  <- (tax_metadata %>%
  filter(grepl("Patescibacteria",Phylum)) %>%
  select(FLASV))$FLASV

Verrucomicrobiota  <- (tax_metadata %>%
  filter(grepl("Verrucomicrobiota",Phylum)) %>%
  select(FLASV))$FLASV

Burkholderiales  <- (tax_metadata %>%
  filter(grepl("Burkholderiales",Order)) %>%
  select(FLASV))$FLASV

Rhizobiales  <- (tax_metadata %>%
  filter(grepl("Rhizobiales",Order)) %>%
  select(FLASV))$FLASV

groupInfo2 <- list(Sphingobacteriales = Sphingobacteriales,
                  Cytophagales = Cytophagales,
                  Acidobacteriota = Acidobacteriota,
                  Myxococcota = Myxococcota,
                  Chloroflexi = Chloroflexi,
                  Cyanobacteria = Cyanobacteria,
                  Planctomycetota = Planctomycetota,
                  Patescibacteria = Patescibacteria,
                  Verrucomicrobiota = Verrucomicrobiota,
                  Burkholderiales = Burkholderiales,
                  Rhizobiales = Rhizobiales)
```


#Create trees
```{r, echo=FALSE, message=FALSE, warning=FALSE}
#load tree
tree1 <- groupOTU(tree_bacteria, groupInfo1)
p1 <- ggtree(tree1, aes(color=group), layout='circular')

tree2 <- groupOTU(tree_bacteria, groupInfo2)
p2 <- ggtree(tree2, aes(color=group), layout='circular')

p <- p1 + p2

ggsave(filename="output/Figure5_Trees.pdf", plot=p, width=20, height=10, useDingbats=FALSE, limitsize=FALSE)
```

